### DISCUSSION TITLE
intermittent 403 response for image requests

### DISCUSSION THREAD

**Discussion by sharkfin009** [directus/directus#17452] [Archive]
### Describe the Bug

I have Cloudinary integrated using the cloudinary storage driver, which is working fine with directus transformations, except when it doesn't , about half the time. 
Something is causing a state where only a plain url without any parameters will work.
Here is my env :

CORS_ENABLED=true
CORS_ORIGIN=true
STORAGE_LOCATIONS=cloudinary
STORAGE_CLOUDINARY_DRIVER=cloudinary
STORAGE_CLOUDINARY_ROOT=
STORAGE_CLOUDINARY_HEALTHCHECK_THRESHOLD=750
STORAGE_CLOUDINARY_CLOUD_NAME=mateas-pics
STORAGE_CLOUDINARY_API_KEY=[key]
STORAGE_CLOUDINARY_API_SECRET=[secret]
STORAGE_CLOUDINARY_ACCESS_MODE=public

As you can see cors is wide open for now, not great but will tighten that up. But there is something else happening and I don't know what.

### To Reproduce

Set up cloudinary acc and configure env as above

### Errors Shown

403 responses

### What version of Directus are you using?

9.22.3

### What version of Node.js are you using?

18.0.0

### What database are you using?

postgres 13

### What browser are you using?

chrome

### How are you deploying Directus?

railway.app

**Comment by sharkfin009** [comment]
Strange thing is this state seems to come and go at random

**Comment by rijkvanzanten** [comment]
What are the error logs in the Directus process? That 403 can be thrown when the requested file can't be found on the storage driver, so it could also be that Cloudinary is preventing the file from being read intermittently

**Comment by sharkfin009** [comment]
thanks Rijk, I see logs of just 403s in the railway deploy logs. Where else can I look? sorry for ignorance

**Comment by rijkvanzanten** [comment]
I'm not quite sure how Railway renders the log output (I don't use it myself), but when Directus gets an unexpected error from the storage provider, it should be logged in stdout before the 403 response is sent to the user. Without those logs, there isn't really any way to know what's going wrong here

**Comment by sharkfin009** [comment]
aha ok i set the log style to raw and heres one of the logs:

```json
{"level":30,"time":1675870894342,"pid":44,"hostname":"railway","req":{"id":111,"method":"GET","url":"/assets/73931747-2af0-4b27-b3e9-c25074488635?key=homepage-dt","query":{"key":"homepage-dt"},"params":{},"headers":{"host":"mateas-directus.up.railway.app","pragma":"no-cache","cache-control":"no-cache","sec-ch-ua":"\"Not_A Brand\";v=\"99\", \"Google Chrome\";v=\"109\", \"Chromium\";v=\"109\"","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36","sec-ch-ua-platform":"\"macOS\"","accept":"image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8","sec-fetch-site":"cross-site","sec-fetch-mode":"no-cors","sec-fetch-dest":"image","referer":"[http://localhost:3001/","accept-encoding":"gzip](http://localhost:3001/%22,%22accept-encoding%22:%22gzip), deflate, br","accept-language":"en-GB,en-US;q=0.9,en;q=0.8","x-forwarded-for":"197.184.182.145","x-forwarded-proto":"https","x-envoy-external-address":"197.184.182.145","x-request-id":"7678e32d-c774-44a2-aa1d-e22d9a8c142d"}},"res":{"statusCode":403,"headers":{"content-security-policy":"default-src none","x-powered-by":"Directus","vary":"Origin","access-control-allow-credentials":"true","access-control-expose-headers":"Content-Range","content-type":"application/json; charset=utf-8","content-length":"102","etag":"W/\"66-OPr7zxcJy7+HqVGdrFe1XpeEIao\""}},"responseTime":668,"msg":"request completed"}
```

**Comment by rijkvanzanten** [comment]
That seems to be the 403 that was returned itself. Are there any other error logs before this line that would indicate a problem in the connection to cloudinary?

**Comment by sharkfin009** [comment]
```

> {
>     "level": 50,
>     "time": 1675870880746,
>     "pid": 44,
>     "hostname": "railway",
>     "err": {
>         "type": "Error",
>         "message": "Location \"cloudinary\" doesn't exist.",
>         "stack": "Error: Location \"cloudinary\" doesn't exist.\n    at StorageManager.location (file:///app/node_modules/@directus/storage/dist/index.js:18:19)\n    at AssetsService.getAsset (/app/node_modules/directus/dist/services/assets.js:75:38)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async /app/node_modules/directus/dist/controllers/assets.js:125:36"
>     },
>     "msg": "Location \"cloudinary\" doesn't exist."
> }

```

**Reply by rijkvanzanten** [reply]
Looks like the `STORAGE_LOCATIONS` aren't properly configured, or maybe the `directus_files.storage` column contains values that are no longer valid (f.e. when the storage locations were reconfigured after files have already been uploaded). This does seem to be a configuration / hosting problem rather than a bug in Directus though, so I'll move this to a discussion 👍🏻

**Reply by ghost** [reply]
I'm having the same issue of receiving `Location \"<STORAGE_LOCATIONS>\" doesn't exist.` when I upgraded to Directus 9.22.4 using the Docker image. I can downgrade to 9.21.2 and it works fine again. Also read [#17286](https://github.com/directus/directus/discussions/17286), but the issue seems like it's not seeing/applying the configuration env var e.g. `STORAGE_LOCATIONS="aws"` and thus ensuing `STORAGE_AWS_*` config.

I'm using `STORAGE_AWS_DRIVER="s3"` rather than `cloudinary`.

**Reply by sharkfin009** [reply]
Thanks guys!
Ok sorry for not looking more carefully at the logs , because it turns out  that the above log was from earlier and not related.  When i do a single image request with a valid parameter, I get just a 403 in the logs like below:

`{
    "level": 30,
    "time": 1675903580638,
    "pid": 43,
    "hostname": "railway",
    "req": {
        "id": 477,
        "method": "GET",
        "url": "/assets/63f2eddd-4b6c-43fb-a0d8-87bcf1906814?key=homepage-dt",
        "query": {
            "key": "homepage-dt"
        },
        "params": {},
        "headers": {
            "host": "mateas-directus.up.railway.app",
            "sec-ch-ua": "\"Not_A Brand\";v=\"99\", \"Google Chrome\";v=\"109\", \"Chromium\";v=\"109\"",
            "sec-ch-ua-mobile": "?0",
            "sec-ch-ua-platform": "\"macOS\"",
            "upgrade-insecure-requests": "1",
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36",
            "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
            "sec-fetch-site": "none",
            "sec-fetch-mode": "navigate",
            "sec-fetch-user": "?1",
            "sec-fetch-dest": "document",
            "accept-encoding": "gzip, deflate, br",
            "accept-language": "en-GB,en-US;q=0.9,en;q=0.8",
            "cookie": "ajs_user_id=e798deff-ca72-43d1-b4d4-510a5555a265; ajs_anonymous_id=7bccc43e-58f6-4cdf-a0cd-08b5743dd8f7; __stripe_mid=d45dd742-8ba9-4612-9ce4-6d624af2f5d17d4e4d; _hp2_ses_props.3707852831=%7B%22ts%22%3A1675903129580%2C%22d%22%3A%22railway.app%22%2C%22h%22%3A%22%2Fproject%2F053b33c8-c0ab-4115-ad2f-30aa18398924%2Fservice%2Fd2b188ab-1f11-4548-8aa6-835577928786%22%7D; directus_refresh_token=TQnyvUTvpDBIfyU8ZvxKuiNsDFr4jai9e-9bG9pNush5YIqnVHDNJ3L6zVXd5_UR; __stripe_sid=084f8136-145c-4e57-9a7e-fdc0b0f2ecc5dfe7f2; _hp2_id.3707852831=%7B%22userId%22%3A%22714211991265351%22%2C%22pageviewId%22%3A%224385799773523922%22%2C%22sessionId%22%3A%228850346974621983%22%2C%22identity%22%3A%22e798deff-ca72-43d1-b4d4-510a5555a265%22%2C%22trackerVersion%22%3A%224.0%22%2C%22identityField%22%3Anull%2C%22isIdentified%22%3A1%7D",
            "x-forwarded-for": "197.184.182.145",
            "x-forwarded-proto": "https",
            "x-envoy-external-address": "197.184.182.145",
            "x-request-id": "5a6a2f67-abc0-4ae8-8a9a-d93f077a52be"
        }
    },
    "res": {
        "statusCode": 403,
        "headers": {
            "content-security-policy": "default-src none",
            "x-powered-by": "Directus",
            "vary": "Origin",
            "access-control-allow-credentials": "true",
            "access-control-expose-headers": "Content-Range",
            "content-type": "application/json; charset=utf-8",
            "content-length": "102",
            "etag": "W/\"66-OPr7zxcJy7+HqVGdrFe1XpeEIao\""
        }
    },
    "responseTime": 308,
    "msg": "request completed"
}`

**Reply by sharkfin009** [reply]
Tried the downgrade to 9.21.2 , that doesnt change things. 

This is an example :
 https://mateas-directus.up.railway.app/assets/63f2eddd-4b6c-43fb-a0d8-87bcf1906814?key=homepage-dt
(returning 403)


Sometimes it returns the image but then back to 403's. At random without me touching the backend... so strange and frustrating. Wonder if it could be a cors issue

**Reply by sharkfin009** [reply]
definitely not CORS, as the issue happens when sec-fetch-mode is navigation also . Not thinking clearly , sorry

**Reply by jbmolle** [reply]
I too have 403 errors. It comes from the Cloudinary Admin API Rate limiting. We are limited at 500 requests / hour for the free version and 2000 requests / hour for the premium.
I think the stat or list function (which use Admin API in Cloudinary Storage) are called too often. 
@rijkvanzanten would you have any ideas here? I can try to work on the code and make a pull request but I don't really know where to start from

**Reply by sharkfin009** [reply]
> I too have 403 errors. It comes from the Cloudinary Admin API Rate limiting. We are limited at 500 requests / hour for the free version and 2000 requests / hour for the premium. I think the stat or list function (which use Admin API in Cloudinary Storage) are called too often. @rijkvanzanten would you have any ideas here? I can try to work on the code and make a pull request but I don't really know where to start from

oh wow @jbmolle  thanks for the insight

**Reply by jbmolle** [reply]
A little update on the problem:
1) Assets service with function getAsset calls storage.location(...).exists(...)
2) Function exists calls stat in Cloudinary Driver
3) stat function uses the Admin API. So if we call more than 500 non cached assets in an hour we can't see our images anymore.
4) In the stat function we just need to get the fields bytes and created_at. We should instead use the explicit method https://cloudinary.com/documentation/image_upload_api_reference#explicit which is not rate limited and sends back the bytes and created_at

**Reply by jbmolle** [reply]
Forgot to link the pull request here #17715
